### ISSUE TITLE
Potential bug in funC optimizer

### ISSUE THREAD

**Issue by bny6613** [ton-blockchain/ton#97]
I have this piece of funC code:
```
(int, int, cell) load_data(cell data){
    var data_cs = data.begin_parse();
    var res = (
        data_cs~load_uint(32), 
        data_cs~load_uint(256), 
        data_cs~load_dict()
    );
    data_cs.end_parse();
    return res;
}

() recv_external(slice in_body) impure {
    var signature = in_body~load_bits(512);
    var cs = in_body;
    var msg_seqno = cs~load_uint(32);
    var (stored_seqno, public_key, dict) = get_data().load_data();
    throw_unless(33, stored_seqno == msg_seqno);
 ...
```
It compiles to :
```
load_data PROC:<{
    //  data
    CTOS	//  data_cs
    32 LDU	//  _4 data_cs
    256 LDU	//  _4 _7 data_cs
    LDDICT	//  res res res data_cs
    ENDS
  }>
  recv_external PROC:<{
    //  in_body
    9 PUSHPOW2	//  in_body _3=512
    LDSLICEX	//  signature in_body
    DUP	//  signature in_body cs
    32 LDU	//  signature in_body msg_seqno cs
    c4 PUSH	//  signature in_body msg_seqno cs _13
    load_data CALLDICT	//  signature in_body msg_seqno cs stored_seqno public_key dict
    s2 s4 PUXC	//  signature in_body dict cs stored_seqno public_key stored_seqno msg_seqno
    EQUAL	//  signature in_body dict cs stored_seqno public_key _16
```
And here are full optimizer rewrites:

```
// *** rewriting s2 PUSH  SWAP  s0 s5 XCHG -> s2 s4 PUXC
// *** rewriting SWAP  s0 s5 XCHG  s4 PUSH -> s0 s5 s4 XC2PU
// *** rewriting 1380271955 PUSHINT  s2 PUSH  SWAP -> OVER  1380271955 PUSHINT
// *** rewriting 1000 PUSHINT  s3 PUSH  SWAP -> s2 PUSH  1000 PUSHINT
// *** rewriting 1000 PUSHINT  s5 PUSH  SWAP -> s4 PUSH  1000 PUSHINT
// *** rewriting 16 PUSHINT  s2 s3 XCHG -> s1 s2 XCHG  16 PUSHINT
// *** rewriting 1347764811 PUSHINT  s2 PUSH  SWAP -> OVER  1347764811 PUSHINT
// *** rewriting DROP  DROP -> 2DROP
// *** rewriting DROP  DROP  DROP -> 3 BLKDROP
// *** rewriting DROP  DROP  DROP  DROP -> 4 BLKDROP 
```

But, if change it and move get_data function inside load_data it produces wrong assembly: 

```
(int, int, cell) load_data(){
    var data_cs = get_data().begin_parse();
    var res = (
        data_cs~load_uint(32), 
        data_cs~load_uint(256), 
        data_cs~load_dict()
    );
    data_cs.end_parse();
    return res;
} 
```
and  
```
var (stored_seqno, public_key, dict) = load_data();
```

which compiles to:

```
load_data PROC:<{
    // 
    c4 PUSH	//  _1
    CTOS	//  data_cs
    32 LDU	//  _4 data_cs
    256 LDU	//  _4 _7 data_cs
    LDDICT	//  res res res data_cs
    ENDS
  }>
  recv_external PROC:<{
    //  in_body
    9 PUSHPOW2	//  in_body _3=512
    LDSLICEX	//  signature in_body
    DUP	//  signature in_body cs
    32 LDU	//  signature in_body msg_seqno cs
    OVER
    load_data CALLDICT
    s0 s5 XCHG	//  signature in_body dict cs stored_seqno public_key stored_seqno msg_seqno
    EQUAL	//  signature in_body dict cs stored_seqno public_key _15
```
Optimizer rewrites:
```
// *** rewriting load_data CALLDICT  s2 PUSH  SWAP -> OVER  load_data CALLDICT
// *** rewriting SWAP  s0 s5 XCHG  s4 PUSH -> s0 s5 s4 XC2PU
// *** rewriting 1380271955 PUSHINT  s2 PUSH  SWAP -> OVER  1380271955 PUSHINT
// *** rewriting 1000 PUSHINT  s3 PUSH  SWAP -> s2 PUSH  1000 PUSHINT
// *** rewriting 1000 PUSHINT  s5 PUSH  SWAP -> s4 PUSH  1000 PUSHINT
// *** rewriting 16 PUSHINT  s2 s3 XCHG -> s1 s2 XCHG  16 PUSHINT
// *** rewriting 1347764811 PUSHINT  s2 PUSH  SWAP -> OVER  1347764811 PUSHINT
// *** rewriting DROP  DROP -> 2DROP
// *** rewriting DROP  DROP  DROP -> 3 BLKDROP
// *** rewriting DROP  DROP  DROP  DROP -> 4 BLKDROP
```
When i execute this code it throws an exception (throw_unless(33, stored_seqno == msg_seqno);) because the stack ordering for EQUAL operation is wrong. It looks like this rewrite `load_data CALLDICT  s2 PUSH  SWAP -> OVER  load_data CALLDICT` messes up the stack and equality check fails. If i disable optimizer (-O0), it works ok.

**Comment by ton-blockchain** [comment]
Yes, this is indeed a bug in the FunC compiler (the result of a CALLDICT is implicitly assumed one stack value wide; this leads to incorrect optimizations). It will be fixed in the next commit.

**Comment by ton-blockchain** [comment]
Fixed in commit f67f5d879ba0a395367a703dcd75f3b87c6ad535. Please check.
